rust/chains/hyperlane-cosmos/src/providers/grpc.rs

```diff
@@ -215,55 +216,61 @@ impl WasmGrpcProvider {
         let amount: u128 = (FixedPointNumber::from(gas_limit) * self.gas_price())
             .ceil_to_integer()
             .try_into()?;
-        let auth_info = signer_info.auth_info(Fee::from_amount_and_gas(
-            Coin::new(
-                // The fee to pay is the gas limit * the gas price
-                amount,
-                self.conf.get_canonical_asset().as_str(),
-            )
-            .map_err(Into::<HyperlaneCosmosError>::into)?,
-            gas_limit,
-        ));
+        let fee_coin = Coin::new(
+            // The fee to pay is the gas limit * the gas price
+            amount,
+            self.conf.get_canonical_asset().as_str(),
+        )
+        .map_err(Into::<HyperlaneCosmosError>::into)?;
+        let auth_info =
+            signer_info.auth_info(Fee::from_amount_and_gas(fee_coin.clone(), gas_limit));
 
         let chain_id = self
             .conf
             .get_chain_id()
             .parse()
             .map_err(Into::<HyperlaneCosmosError>::into)?;
 
-        Ok(
+        Ok((
             SignDoc::new(&tx_body, &auth_info, &chain_id, account_info.account_number)
                 .map_err(Into::<HyperlaneCosmosError>::into)?,
-        )
+            fee_coin,
+        ))
     }
 
-    /// Generates a raw signed transaction including `msgs`, estimating gas if a limit is not provided.
-    async fn generate_raw_signed_tx(
+    /// Generates a raw signed transaction including `msgs`, estimating gas if a limit is not provided,
+    /// and the Coin amount required to pay for tx fees.
+    async fn generate_raw_signed_tx_and_fee(
         &self,
         msgs: Vec<cosmrs::Any>,
         gas_limit: Option<u64>,
-    ) -> ChainResult<Vec<u8>> {
+    ) -> ChainResult<(Vec<u8>, Coin)> {
         let gas_limit = if let Some(l) = gas_limit {
             l
         } else {
             self.estimate_gas(msgs.clone()).await?
         };
 
-        let sign_doc = self.generate_unsigned_sign_doc(msgs, gas_limit).await?;
+        let (sign_doc, fee) = self
+            .generate_unsigned_sign_doc_and_fee(msgs, gas_limit)
+            .await?;
 
         let signer = self.get_signer()?;
         let tx_signed = sign_doc
             .sign(&signer.signing_key()?)
             .map_err(Into::<HyperlaneCosmosError>::into)?;
-        Ok(tx_signed
-            .to_bytes()
-            .map_err(Into::<HyperlaneCosmosError>::into)?)
+        Ok((
+            tx_signed
+                .to_bytes()
+                .map_err(Into::<HyperlaneCosmosError>::into)?,
+            fee,
+        ))
     }
 
     /// Estimates gas for a transaction containing `msgs`.
     async fn estimate_gas(&self, msgs: Vec<cosmrs::Any>) -> ChainResult<u64> {
```

Comment:
do you happen to have a log I can see when this happened?

I'm surprised it would error because when we estimate gas we are setting the gas limit to 0, which would mean that the account pays 0 fees

Reply:
because my fix just compares what the account would spend on fees to the balance, if we're getting errors even with 0 fees im worried my fix doesn't cover all cases